### PR TITLE
Convert temporal column and refresh period to datafusion expr

### LINKED ISSUE #1175: Generate sql filter based on dataset schema and temporal column and format for refresh period

### PR DESCRIPTION
**Pull request by y-f-u**
Add a timestamp converter to simplify the logic.
Convert the refresh period logic to expr.

There will be following PR to enable the filter when fetching the data.

Fixes #1175 

Edit:
With the merge of #1188 . This allows the period to work properly

```
  - from: s3://spiceai-demo-datasets/taxi_trips/2024/
    name: taxi_trips
    time_column: tpep_pickup_datetime
    acceleration:
      enabled: true
      refresh_interval: 10m
      refresh_period: 1000w
      refresh_sql: |
        SELECT * FROM taxi_trips WHERE passenger_count = 1;
 ```

### CODE REVIEW

File: crates/runtime/src/accelerated_table.rs [hunk 1]
```diff
@@ -400,64 +427,91 @@ impl AcceleratedTable {
...
-                    if time_format == TimeFormat::UnixMillis {
-                        scale = 1000;
+#[derive(Clone, Debug)]
+struct TimestampFilterConvert {
+    time_column: String,
+    time_format: ExprTimeFormat,
+}
+
+#[allow(clippy::needless_pass_by_value)]
+impl TimestampFilterConvert {
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs]: This seems unrelated to the accelerated_table - let's put this in a separate module

File: crates/runtime/src/accelerated_table.rs [hunk 2]
```diff
@@ -400,64 +427,91 @@ impl AcceleratedTable {
...
+    fn create(
+        field: Option<(usize, &arrow::datatypes::Field)>,
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs]: Seems like the index information in the tuple isn't used? `(usize, &arrow::datatypes::Field)` it would be simpler just to pass the arrow field here and strip the index.

File: crates/runtime/src/accelerated_table.rs [hunk 3]
```diff
@@ -372,16 +371,44 @@ impl AcceleratedTable {
...
+
                     let mut refresh_stream = ReceiverStream::new(receiver);
 
                     while refresh_stream.next().await.is_some() {
-                        tracing::info!("Refreshing data for {dataset_name}");
+                        tracing::info!("[refresh] Refreshing data for {dataset_name}");
                         status::update_dataset(&dataset_name, status::ComponentStatus::Refreshing);
                         let timer = TimeMeasurement::new("load_dataset_duration_ms", vec![("dataset", dataset_name.clone())]);
-                        let all_data = match get_all_data(&mut ctx, OwnedTableReference::bare(dataset_name.clone()), Arc::clone(&federated), refresh_sql.clone()).await {
+                        let  filters = match (refresh.period, filter_converter.clone()){
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs]:
```suggestion
                        let filters = match (refresh.period, filter_converter.clone()){
```

File: crates/runtime/src/accelerated_table.rs [hunk 4]
```diff
@@ -372,16 +371,44 @@ impl AcceleratedTable {
...
+                        (Some(_), None) => {
+                            tracing::warn!("[refresh] No time_column is provided, refresh_period will be ignored");
+                            None
+                        }
+                        (Some(_), Some(column)) => {
+                            let schema = federated.schema();
+                            let field = schema.column_with_name(column.as_str());
+                            TimestampFilterConvert::create(field, Some(column), refresh.time_format)
+                        }
+                    };
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs]: This feels unnecessarily complex. Is there a simpler way to achieve what we want?

File: crates/runtime/src/accelerated_table.rs [hunk 5]
```diff
@@ -372,16 +371,44 @@ impl AcceleratedTable {
...
+                                let Ok(timestamp) = get_timestamp(start) else {
+                                    tracing::error!("[refresh] Failed to get timestamp");
+                                    continue;
+                                };
+                                vec![converter.convert(timestamp, Operator::Gt)]
+                            },
+                            _ => vec![],
+                        };
+
+                        let all_data = match get_all_data(&mut ctx, OwnedTableReference::bare(dataset_name.clone()), Arc::clone(&federated), refresh.sql.clone(), filters).await {
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs]:
```suggestion
                        let data = match get_data(&mut ctx, OwnedTableReference::bare(dataset_name.clone()), Arc::clone(&federated), refresh.sql.clone(), filters).await {
```

This isn't technically all data anymore.

File: crates/runtime/src/accelerated_table.rs [hunk 6]
```diff
@@ -400,64 +427,91 @@ impl AcceleratedTable {
...
+        Some(Self {
+            time_column,
+            time_format,
+        })
+    }
+
+    #[allow(clippy::cast_possible_wrap)]
+    fn convert(&self, timestamp: u64, op: Operator) -> Expr {
+        let format = self.time_format.clone();
+        let time_column: &str = &self.time_column.clone();
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs:489]:
```suggestion
        let time_column: &str = self.time_column.as_str;
```

File: crates/runtime/src/accelerated_table.rs [hunk 7]
```diff
@@ -400,64 +427,91 @@ impl AcceleratedTable {
...
+        let expr_time_format = format.clone();
+        match expr_time_format {
```
Comment (phillipleblanc) [inline crates/runtime/src/accelerated_table.rs]:
```suggestion
        let time_column: &str = self.time_column.as_str();
        match &self.time_format {
```

File: crates/runtime/src/accelerated_table.rs (hunk 5 above)
Comment (y-f-u) [reply crates/runtime/src/accelerated_table.rs]: good point.

File: crates/runtime/src/accelerated_table.rs (hunk 4 above)
Comment (y-f-u) [reply crates/runtime/src/accelerated_table.rs]: it might be. The main complication comes from informative information to the user if they have set `refresh_period` and expect it is working.

File: crates/runtime/src/accelerated_table.rs (hunk 3 above)
Comment (y-f-u) [reply crates/runtime/src/accelerated_table.rs]: Surprised clippy was accepting that.

File: crates/runtime/src/accelerated_table.rs (hunk 2 above)
Comment (y-f-u) [reply crates/runtime/src/accelerated_table.rs]: Yeah It is not used. It was mainly for not doing extra logic to split the tuple in the caller spaces as there are two callers. And the tuple kinda make sense being together.

I had the other version too before this and opted for this one.

File: crates/runtime/src/accelerated_table.rs (hunk 4 above)
Comment (phillipleblanc) [reply crates/runtime/src/accelerated_table.rs]: That type of validation/warning should happen when we are loading the spicepod imo

Comment (y-f-u) [reply crates/runtime/src/accelerated_table.rs]: moved to `create`

**Review by sgrebnov — DISMISSED** [review]

File: crates/runtime/src/datafusion/filter_converter.rs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        Some(Self {
+            time_column,
+            time_format,
+        })
+    }
+
+    #[allow(clippy::cast_possible_wrap)]
+    pub(crate) fn convert(&self, timestamp: u64, op: Operator) -> Expr {
+        let time_column: &str = &self.time_column.clone();
+        match self.time_format.clone() {
```
Comment (phillipleblanc) [inline crates/runtime/src/datafusion/filter_converter.rs]:
```suggestion
        match &self.time_format {
```

File: crates/runtime/src/datafusion/filter_converter.rs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+        Some(Self {
+            time_column,
+            time_format,
+        })
+    }
+
+    #[allow(clippy::cast_possible_wrap)]
+    pub(crate) fn convert(&self, timestamp: u64, op: Operator) -> Expr {
+        let time_column: &str = &self.time_column.clone();
```
Comment (phillipleblanc) [inline crates/runtime/src/datafusion/filter_converter.rs]:
```suggestion
        let time_column: &str = self.time_column.as_ref();
```

File: crates/runtime/src/datafusion/filter_converter.rs (hunk 8 above)
Comment (phillipleblanc) [reply crates/runtime/src/datafusion/filter_converter.rs]: I pushed a commit for this

File: crates/runtime/src/datafusion/filter_converter.rs (hunk 9 above)
Comment (phillipleblanc) [reply crates/runtime/src/datafusion/filter_converter.rs]: I pushed a commit for this

**Review by phillipleblanc — APPROVED** [review]
